pyomo/util/subsystems.py

```diff
@@ -0,0 +1,195 @@
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright 2017 National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+from pyomo.core.base.PyomoModel import ConcreteModel
+from pyomo.core.base.reference import Reference
+from pyomo.core.expr.visitor import identify_variables
+from pyomo.common.collections import ComponentSet, ComponentMap
+
+
+def create_subsystem_block(constraints, variables=None, include_fixed=False):
+    """ This function defines creates a block to serve as a subsystem with
+    the specified variables and constraints. To satisfy certain writers,
+    other variables that appear in the constraints must be added to the block
+    as well. We call these the "input vars." They may be thought of as
+    parameters in the subsystem, but we do not fix them here as it is not
+    obvious that this is desired.
+    """
+    if variables is None:
+        variables = []
+    block = ConcreteModel()
+    block.vars = Reference(variables)
+    block.cons = Reference(constraints)
+    var_set = ComponentSet(variables)
+    input_vars = []
+    for con in constraints:
+        for var in identify_variables(con.body, include_fixed=include_fixed):
+            if var not in var_set:
+                input_vars.append(var)
+                var_set.add(var)
+    block.input_vars = Reference(input_vars)
+    return block
+
+
+def generate_subsystem_blocks(subsystems, include_fixed=False, fix_inputs=True):
+    """ Generates blocks that contain subsystems of variables and constraints.
+    
+    Arguments
+    ---------
+    subsystems: List of tuples. Each tuple is a list of constraints then
+                a list of variables that will define a subsystem.
+    include_fixed: Whether to add already fixed variables to the generated
+                   subsystem blocks.
+    fix_inputs: Whether to fix variables in the constraints that are not
+                included in the subsystem.
+
+    Yields
+    ------
+    "Subsystem blocks" containing the variables and constraints specified
+    by each entry in subsystems. Variables in the constraints that are
+    not specified are contained in the input_vars component.
+    """
+    for cons, vars in subsystems:
+        block = create_subsystem_block(cons, vars, include_fixed)
+        if fix_inputs:
+            to_fix = list(block.input_vars[:])
+            with TemporarySubsystemManager(to_fix=to_fix):
+                # On enter, we fix the "inputs" of the subsystem. When
+                # control returns to this code (iterator advance), we
+                # exit this context, and inputs are unfixed.
+                yield block
```

Comment:
I like this: it looks pretty clean.  However, is this guaranteed to unfix the block if the iterator is not exhausted?  For example, if an exception is raised by a solver while iterating over `generate_subsystem_blocks`, is the last block solved unfixed?

Reply:
I believe so. `__exit__` is called before the program stops or the exception is handled. I have added a test to verify this behavior.

Reply:
There appears to be a pypy-only failure of exactly this test. Will try to debug tomorrow.

Reply:
It appears that having the context manager inside of the generator function leads to implementation-specific behavior if an exception is raised. Since this can lead to an implementation-dependent state of the user's model as they process the exception, I removed the embedded `TemporarySubsystemManager` from the generator. If the user wants to fix the inputs to a subsystem block, they can add the context manager themselves.